evm/src/arithmetic/modular.rs

```diff
@@ -573,15 +643,15 @@ mod tests {
                     }
                 }
 
-                generate(&mut lv, op_filter);
+                generate(&mut lv, &mut nv, op_filter);
 
                 let mut constraint_consumer = ConstraintConsumer::new(
                     vec![GoldilocksField(2), GoldilocksField(3), GoldilocksField(5)],
                     GoldilocksField::ONE,
                     GoldilocksField::ONE,
                     GoldilocksField::ONE,
                 );
```

Comment:
The arguments of `ConstraintConsumer::new` are `alphas: Vec<P::Scalar>, z_last: P, lagrange_basis_first: P, lagrange_basis_last: P`. My understanding (please confirm, @wborgeaud) is that `z_last` is a filter to enable `constraint_transition`, `lagrange_basis_first` enables `constraint_first_row`, and `lagrange_basis_last` enables `constraint_last_row`.

We have a constraint to ensure that `lv` is not the last row when performing modular operations, so for tests to pass `constraint_last_row` must be disabled (i.e. `lagrange_basis_last == GoldilocksField::ZERO`).

```suggestion
                let mut constraint_consumer = ConstraintConsumer::new(
                    vec![GoldilocksField(2), GoldilocksField(3), GoldilocksField(5)],
                    GoldilocksField::ONE,
                    GoldilocksField::ZERO,
                    GoldilocksField::ZERO,
                );
```

The same applies to the `zero_modulus` test.

Reply:
Yes that's correct. If you are trying to emulate a row that is not the last row in the trace, you should set `z_last` to a non-zero value and `lagrange_basis_last` to zero.
Also, small nit but I think it would be better to use random values for the `alphas` here to make the test sounder.

Reply:
Thanks both! In @nbgl's suggested change, both `lagrange_basis_first` and `lagrange_basis_last` are set to zero, but in the explanations you both provided it seems like only `lagrange_basis_last` should be set to zero. Was `lagrange_basis_first` set to zero by accident in the change or is there something I've missed? (The tests do pass when `lagrange_basis_first` is left equal to one.)

Reply:
`lagrange_basis_first` should be set to one iff you are emulating the first row of the trace. I'm guessing you don't have any logic depending on whether a row is the first (i.e., you don't use `consumer.constraint_first_row`), so it shouldn't matter if you set `lagrange_basis_first` to zero or one.